fix: workaround leak through promise resolvers

mhofman opened the pull request:
The investigation into https://github.com/Agoric/agoric-sdk/issues/5507 lead us into discovering a leak caused by promises, in particular when using [`Promise.race` with a long pending promise](https://github.com/nodejs/node/issues/17469). The root cause really seem that the promise resolve functions hold onto the promise strongly (instead of just an internal record representing the reactions registered on the promise), and through that, the result of said promise. `Promise.race` is implemented as adding the result's promise's resolver functions to each of the contenders reactions, causing the long pending promise to keep the raced result alive.

This PR re-implements `Promise.race` based on the solution in https://github.com/nodejs/node/issues/17469#issuecomment-685216777. It also updates `makePromiseKit` to return resolvers that will sever their hold on the original resolvers (and thus the promise) once called. That latter approach, while simpler, is not sufficient for `Promise.race` as reactions would keep queueing on the long pending promise (engines are not sufficiently smart to realize a reaction is a native resolver that has become inert).

Finally it installs the new race implementation as a `Promise.race` vetted shim in `@endo/init`.

(I'm half considering re-writing promises in userland with proper support for short-circuiting, which would avoid most of these issues, but the few spec places creating primordial promises would damper that. A long term solution might be to rewrite the spec  with proper short-circuiting, maybe as part of [tc39/proposal-faster-promise-adoption](https://github.com/tc39/proposal-faster-promise-adoption))

kriskowal reviewed (approved):
We could alternately move this into `ses` as a taming of the existing `Promise.race`. It would only be a slight feature creep.

michaelfig commented:
> We could alternately move this into `ses` as a taming of the existing `Promise.race`. It would only be a slight feature creep.

Why not just as a separate "vetted shim" package, loaded by `@endo/init`?

kriskowal commented:
> > We could alternately move this into `ses` as a taming of the existing `Promise.race`. It would only be a slight feature creep.
> 
> Why not just as a separate "vetted shim" package, loaded by `@endo/init`?

That’s even better.

mhofman commented:
Updated to use a vetted shim approach, and reverted the explicit uses to rely on the shim.

Also updated the implementation to follow the spec more closely. It'd be nice to test against test262, but I don't believe we have the infra for that.

mhofman commented on packages/promise-kit/shim.js:
```diff
@@ -0,0 +1,4 @@
+import { memoRace } from './memo-race.js';
+
+// Unconditionally replace with a non-leaking version
+Promise.race = memoRace;
```
It is and is one of the thing I fixed in https://github.com/endojs/endo/pull/1222/commits/8cfe2a1b106468402f11e00c079987b0ed5e0f61 to preserve the behavior.

erights commented on packages/promise-kit/index.js:
```diff
@@ -3,17 +3,26 @@
 
+/**
+ * @template T
+ * @callback PromiseExecutor The promise executor
```
How aware is the promiseKit package of eventual-send / HandledPromise? Should the executor type include `resolveToPresence`?

mhofman replied:
It is not and ignoring it, like the existing `makePromiseKit`

erights commented on packages/promise-kit/memo-race.js:
```diff
@@ -0,0 +1,154 @@
+    let deferred;
+    /** @type {T[]} */
+    const cachedValues = [];
+    const C = this || Promise;
```
Yeah, as a shim, I think we should be as strict as possible about spec compliance. Thanks.

erights commented on packages/promise-kit/index.js:
```diff
@@ -3,17 +3,26 @@
 
 /// <reference types="ses"/>
```
Now that index.js is growing and has other *.js peers, we need to follow our normal convention of having index.js just be a pass-through, and have all the substantive code in non-index `src/*.js` files.

kriskowal replied:
In my opinion, this is still small enough not to need additional ceremony. Both of the modules at this level need entry points in `package.json` because they are exposed API.

kriskowal replied:
Which is to say, moving code into `src`  would not eliminate these two files, but hollow both out and leave `export * from './src/x.js` ceremony.

mhofman replied:
I just took a stab at splitting into different files and it's not straightforward, mainly because of the types intermingling and the `BestPipelinablePromise` stuff. Best I could do is move the promise executor stuff into `src`, maybe even the `isPromise` too, but leave the others behind.

erights reviewed (changes requested):
Not done yet. Just giving you bits as they occur to minimize latency.

erights commented on packages/promise-kit/src/promise-executor-kit.js:
```diff
@@ -0,0 +1,56 @@
+ * @template T
+ * @returns {Pick<import('./types.js').PromiseKit<T>, 'resolve' | 'reject'> & { executor: PromiseExecutor<T>}}
+ */
+export const makeReleasingExecutorKit = () => {
```
Elegant!

erights commented on packages/promise-kit/src/memo-race.js:
```diff
@@ -0,0 +1,154 @@
+/**
+ * @template [T=any]
+ * @typedef {object} Deferred
+ * @property {(value?: T | PromiseLike<T> ) => void} resolve
```
Are you trying to stay close to the original source, or trying to adapt to our style? If the first, ignore this suggestion.

```suggestion
 * @property {(value?: ERef<T> ) => void} resolve
```

Also import `isObject` rather than redefine it?

mhofman replied:
I was trying to stay close to the original, but at this point, I've made so many changes this would be warranted.

Regarding `isObject`, it's actually not exported by any package, at least not any that would make sense taking a dependency on.

erights reviewed (approved):
I did not do a thorough review of memo-race.js . But nothing seemed wrong, and you already have @kriskowal 's approval. I'll defer to that. Please proceed.

mhofman commented:
Added a blurb to `NEWS.md` and addressed @erights type feedback in https://github.com/endojs/endo/commit/0377da065fd1af4012797121c6f59f9223b86f40 then rebase the whole thing cleaning up the history: [0377da0..219e49d](https://github.com/endojs/endo/compare/0377da065fd1af4012797121c6f59f9223b86f40..219e49dcdb61bddad9bb6ff8662aed38b6d89b8e)
